lib/attack_test.go

```diff
@@ -267,6 +270,59 @@ func TestMaxBody(t *testing.T) {
 	}
 }
 
+func TestUnixSocket(t *testing.T) {
+	t.Parallel()
+	body := []byte("IT'S A UNIX SYSTEM, I KNOW THIS")
+
+	socketDir, err := ioutil.TempDir("", "vegata")
+	if err != nil {
+		t.Fatal("Failed to create socket dir", err)
+	}
+	defer os.RemoveAll(socketDir)
+	socketFile := filepath.Join(socketDir, "test.sock")
+
+	unixListener, err := net.Listen("unix", socketFile)
+
+	if err != nil {
+		t.Fatal("Failed to listen on unix socket", err)
+	}
+
+	server := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write(body)
+		}),
+	}
+	defer server.Close()
+
+	go func() {
+		server.Serve(unixListener)
+	}()
+
+	start := time.Now()
+	for {
+		if time.Since(start) > 1*time.Second {
+			t.Fatal("Server didn't listen on  unix socket in time")
+		}
+		_, err := os.Stat(socketFile)
+		if err == nil {
+			break
+		} else if os.IsNotExist(err) {
+			time.Sleep(10 * time.Millisecond)
+		} else {
+			t.Fatalf("unexpected error from unix socket %s", err)
```

Comment:
Calling `t.Fatal` will prevent defers from running. If you want defers to run, call `t.Error` followed by return.

Reply:
Fatal calls FailNow. FailNow calls runtime.Goexit, and Goexit causes defers to run.

Reply:
@dominikh: Thanks!! This is great. I am not sure how I came to conclude that t.Fatal didn't run defers. Now reading the docs of t.FailNow makes that very clear.

Reply:
Fixed: https://github.com/tsenart/vegeta/commit/7bf09f4fab4d852141935796455c17459d212098